### PR TITLE
Add optional ability to have persistent csrf token

### PR DESCRIPTION
**Pull request by infernalmaster**
Related to #467

### CODE REVIEW

**Comment by infernalmaster** [comment]
[First time](https://github.com/amberframework/amber/pull/474/commits/de2552678709ab11c87565e421483f505646375e) I implemented that simple way. But than I looked to Rails sources and found that [comment](https://github.com/rails/rails/blob/5.1.3/actionpack/lib/action_controller/metal/request_forgery_protection.rb#L294)

**Comment by elorest** [comment]
So the general idea is that the token can't be the same every time or someone could guess the ssl encryption key after multiple requests. Similar to padding oracle.

Thanks.

File: src/amber/router/pipe/csrf.cr [hunk 1]
```diff
@@ -22,20 +25,89 @@ module Amber
...
+        module TokenOperations
+          extend self
+
+          # Creates a masked version of the authenticity token that varies
+          # on each request. The masking is used to mitigate SSL attacks
+          # like BREACH.
+          def mask(unmasked_token)
+            one_time_pad = Bytes.new(TOKEN_LENGTH).tap { |buf| Random::Secure.random_bytes(buf) }
+            encrypted_csrf_token = xor_byte_strings(one_time_pad, unmasked_token.bytes)
+            "#{String.new(one_time_pad)}#{encrypted_csrf_token}"
```
Comment (elorest) [inline src/amber/router/pipe/csrf.cr]: String.new is problematic as the bytes contained in `one_time_pad` aren't all ascii chars. We'll need to base64 encode it first.

File: src/amber/router/pipe/csrf.cr [hunk 2]
```diff
@@ -22,20 +25,89 @@ module Amber
...
+
+          def unmask(masked_token)
+            one_time_pad = masked_token.bytes[0...TOKEN_LENGTH]
+            encrypted_csrf_token = masked_token.bytes[TOKEN_LENGTH..-1]
+            xor_byte_strings(one_time_pad, encrypted_csrf_token)
+          end
+
+          def xor_byte_strings(s1, s2) : String
+            s1.each_with_index { |c1, i| s2[i] ^= c1 }
+            String.new(s2.to_unsafe, TOKEN_LENGTH)
```
Comment (elorest) [inline src/amber/router/pipe/csrf.cr]: Same here with String.new.

File: src/amber/router/pipe/csrf.cr [hunk 3]
```diff
@@ -22,20 +25,89 @@ module Amber
...
+            "#{String.new(one_time_pad)}#{encrypted_csrf_token}"
+          end
+
+          def unmask(masked_token)
+            one_time_pad = masked_token.bytes[0...TOKEN_LENGTH]
+            encrypted_csrf_token = masked_token.bytes[TOKEN_LENGTH..-1]
+            xor_byte_strings(one_time_pad, encrypted_csrf_token)
+          end
+
+          def xor_byte_strings(s1, s2) : String
```
Comment (elorest) [inline src/amber/router/pipe/csrf.cr]: These are bytes not strings. I would suggest returning bytes as well, then you can base64 the entire thing once.

**Comment by infernalmaster** [comment]
Sorry for late reply. I did some changes, I'm not sure if it's what you wanted. This is the best I can do according to my knowledges of Crystal.

File: src/amber/router/pipe/csrf.cr [hunk 4]
```diff
@@ -22,20 +25,94 @@ module Amber
...
+      module PersistentToken
+        extend self
+        extend BaseToken
+
+        def valid_token?(context)
+          request_token = request_token(context)
+          return false if request_token.nil?
+          decoded_request_token = Base64.decode_string(request_token)
+          return false if decoded_request_token.bytesize != 2 * TOKEN_LENGTH
+          unmaseked_request_token = TokenOperations.unmask(decoded_request_token)
```
Comment (elorest) [inline src/amber/router/pipe/csrf.cr]: Typo.

File: src/amber/router/pipe/csrf.cr [hunk 5]
```diff
@@ -8,10 +9,12 @@ module Amber
       HEADER_KEY    = "X-CSRF-TOKEN"
       PARAM_KEY     = "_csrf"
       CSRF_KEY      = "csrf.token"
+      TOKEN_LENGTH  = 32
+
+      class_property token_strategy : PersistentToken | RefreshableToken = PersistentToken
```
Comment (elorest) [inline src/amber/router/pipe/csrf.cr]: Has anyone benchmarked the difference between these two?

**Comment by eliasjpr** [comment]
@elorest can you address your comments in a commit so we can merge it in? WDYT?

**Comment by eliasjpr** [comment]
Issue https://github.com/amberframework/amber/issues/467

**Comment by infernalmaster** [comment]
For benchmarks I just wrote this code inside test cases 1:

```
          csrf = CSRF.new
          request = HTTP::Request.new("PUT", "/")
          context = create_context(request)

          CSRF.token_strategy = Amber::Pipe::CSRF::RefreshableToken
          Benchmark.ips do |x|
            x.report("RefreshableToken") do
              token = CSRF.token(context)
              context.params[Amber::Pipe::CSRF::PARAM_KEY] = token.to_s
              result = csrf.call(context)
              if result != nil
                raise Exception.new("csrf failed")
              end
            end
          end

          CSRF.token_strategy = Amber::Pipe::CSRF::PersistentToken
          Benchmark.ips do |x|
            x.report("PersistentToken") do
              token = CSRF.token(context)
              context.params[Amber::Pipe::CSRF::PARAM_KEY] = token.to_s
              result = csrf.call(context)
              if result != nil
                raise Exception.new("csrf failed")
              end
            end
          end
```

Than I used flag --release and results are:
```
RefreshableToken 380.88k (  2.63µs) (± 4.40%) fastest
PersistentToken 308.93k (  3.24µs) (± 6.16%) fastest
```

I also tried almost the same (test case 2):

```

          n = _____
          Benchmark.bm do |x|
            x.report("RefreshableToken:") do
              CSRF.token_strategy = Amber::Pipe::CSRF::RefreshableToken
              csrf = CSRF.new
              request = HTTP::Request.new("PUT", "/")
              context = create_context(request)

              n.times do
                token = CSRF.token(context)
                context.params[Amber::Pipe::CSRF::PARAM_KEY] = token.to_s
                result = csrf.call(context)
                if result != nil
                  raise Exception.new("csrf failed")
                end
              end
            end
            x.report("PersistentToken:") do
              CSRF.token_strategy = Amber::Pipe::CSRF::PersistentToken
              csrf = CSRF.new
              request = HTTP::Request.new("PUT", "/")
              context = create_context(request)

              n.times do
                token = CSRF.token(context)
                context.params[Amber::Pipe::CSRF::PARAM_KEY] = token.to_s
                result = csrf.call(context)
                if result != nil
                  raise Exception.new("csrf failed")
                end
              end
            end
          end
```

With n=100_000:

```
RefreshableToken:   0.050000   0.230000   0.280000 (  0.265488)
PersistentToken:    0.100000   0.240000   0.340000 (  0.313201)
```

N=1000

```
RefreshableToken:   0.000000   0.000000   0.000000 (  0.002422)
PersistentToken:    0.010000   0.010000   0.020000 (  0.004350)
```


Than I tried simple test (test case 3):

```
          n = 100000
          Benchmark.bm do |x|
            x.report("RefreshableToken:") do
              CSRF.token_strategy = Amber::Pipe::CSRF::RefreshableToken

              n.times do
                t = CSRF.token(context)
                if t == "x"
                  raise Exception.new("csrf failed")
                end
              end
            end
            x.report("PersistentToken:") do
              CSRF.token_strategy = Amber::Pipe::CSRF::PersistentToken

              n.times do
                t = CSRF.token(context)
                if t == "x"
                  raise Exception.new("csrf failed")
                end
              end
            end
          end
```
Results:
```
RefreshableToken:   0.010000   0.000000   0.010000 (  0.002404)
PersistentToken:    0.080000   0.270000   0.350000 (  0.298745)
```

So I tried to understand why previous results where close for this strategies and tried to run test case 2 (n=100_000) without this operations `context.session.delete(CSRF_KEY)` inside RefreshableToken:

```
RefreshableToken:   0.020000   0.000000   0.020000 (  0.022798)
PersistentToken:    0.100000   0.250000   0.350000 (  0.318500)
```

So looks like session modification is very slow: 0.265488 - 0.022798 = 0.24269 seconds.

**Pull request merged**